cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,282 @@
+    }
+
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = checkJobQueue();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Check jobs in job queue for constraint satisfaction.
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long checkJobQueue() {
+      boolean emptyFetch = false;
+      try {
+        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            return checkJobConstraints(context);
+          }
+        });
+        failureCount = 0;
+      } catch (Exception e) {
+        LOG.warn("Failed to check Job constraints. Will retry in next run", e);
+        failureCount++;
+      }
+
+      try {
+        Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            // run any ready jobs
+            return runReadyJobs(context);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to launch programs. Will retry in next run", e);
+        failureCount++;
+      }
+
+      // If there is any failure, delay the next fetch based on the strategy
+      if (failureCount > 0) {
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+
+      // Sleep for 2 seconds if there's no jobs in the queue
+      return emptyFetch ? 2000L : 0L;
+    }
+
+    private boolean checkJobConstraints(DatasetContext context) throws Exception {
+      boolean emptyScan = true;
+
+      JobQueueDataset jobQueue = getJobQueue(context);
+      try (CloseableIterator<Job> jobsIter = jobQueue.getAllJobs()) {
+        while (jobsIter.hasNext() && isRunning()) {
+          emptyScan = false;
+          Job job = jobsIter.next();
+          switch (job.getState()) {
+            case PENDING_TRIGGER:
+              if (isTriggerSatisfied(job)) {
+                job.setState(Job.State.PENDING_CONSTRAINT);
+                jobQueue.put(job);
+              }
+              break;
+            case PENDING_CONSTRAINT:
+              if (constraintsSatisfied(job)) {
+                job.setState(Job.State.PENDING_LAUNCH);
+                jobQueue.put(job);
+              }
+              break;
+          }
+        }
+      }
+      return emptyScan;
+    }
+
+    private boolean runReadyJobs(DatasetContext context) throws IOException, DatasetManagementException {
```

Comment:
can we discuss the algorithm behind this? If I understand correctly, you do this:

1. full scan over all jobs to check constraints and mark the ones that are ready
2. another full scan over all jobs to launch the ready ones

I am not sure that is the best strategy. If you have a thousand jobs, it may take 100 seconds to check all constraints (or longer). That adds latency to the first job that can be launched.